### PR TITLE
Add execution revert message in receipt 

### PR DESCRIPTION
**Pull request by koseoyoung**
works on #2530

### CODE REVIEW

File: action/protocol/execution/evm/evm.go [hunk 1]
```diff
@@ -233,6 +238,16 @@ func ExecuteContract(
 		hu.IsPre(config.Greenland, blkCtx.BlockHeight) && receipt.Status == uint64(iotextypes.ReceiptStatus_ErrCodeStoreOutOfGas) {
 		receipt.AddTransactionLogs(stateDB.TransactionLogs()...)
 	}
+
+	if hu.IsPost(config.Hawaii, blkCtx.BlockHeight) && receipt.Status == uint64(iotextypes.ReceiptStatus_ErrExecutionReverted) {
+		if retval != nil && bytes.Equal(retval[:4], revertSelector) {
+			// in case of the execution revert error, parse the retVal and add to receipt
+			data := retval[4:]
+			msgLength := byteutil.BytesToUint64BigEndian(data[56:64])
```
Comment (dustinxie) [inline action/protocol/execution/evm/evm.go:246]: what's in the first 32-bytes data[:32]?

File: action/protocol/execution/evm/evmstatedbadapter.go [hunk 2]
```diff
@@ -21,14 +21,13 @@ import (
...
 	"github.com/iotexproject/iotex-address/address"
-	"github.com/iotexproject/iotex-proto/golang/iotextypes"
-
 	"github.com/iotexproject/iotex-core/action"
 	"github.com/iotexproject/iotex-core/action/protocol"
 	accountutil "github.com/iotexproject/iotex-core/action/protocol/account/util"
 	"github.com/iotexproject/iotex-core/db/trie"
 	"github.com/iotexproject/iotex-core/pkg/log"
 	"github.com/iotexproject/iotex-core/state"
+	"github.com/iotexproject/iotex-proto/golang/iotextypes"
```
Comment (dustinxie) [inline action/protocol/execution/evm/evmstatedbadapter.go:30]: no need to move here, iotex-proto is outside of iotex-core and considered same as iotex-address, go-pkgs

File: action/protocol/execution/protocol_test.go [hunk 3]
```diff
@@ -45,6 +43,7 @@ import (
 	"github.com/iotexproject/iotex-core/state/factory"
 	"github.com/iotexproject/iotex-core/test/identityset"
 	"github.com/iotexproject/iotex-core/testutil"
+	"github.com/iotexproject/iotex-proto/golang/iotextypes"
```
Comment (dustinxie) [inline action/protocol/execution/protocol_test.go:46]: not needed

File: action/protocol/execution/evm/evm.go (hunk 1 above)
Comment (koseoyoung) [reply action/protocol/execution/evm/evm.go:246]:
```
0x08c379a0                                                         // Function selector for Error(string)
0x0000000000000000000000000000000000000000000000000000000000000020 // Data offset
0x000000000000000000000000000000000000000000000000000000000000001a // String length
0x4e6f7420656e6f7567682045746865722070726f76696465642e000000000000 // String data
```

I thought it is calldata offset

File: action/receipt.go [hunk 4]
```diff
@@ -13,9 +13,8 @@ import (
 
 	"github.com/iotexproject/go-pkgs/hash"
 	"github.com/iotexproject/iotex-address/address"
-	"github.com/iotexproject/iotex-proto/golang/iotextypes"
-
 	"github.com/iotexproject/iotex-core/pkg/log"
+	"github.com/iotexproject/iotex-proto/golang/iotextypes"
```
Comment (dustinxie) [inline action/receipt.go:17]: no need to move

File: action/receipt.go [hunk 5]
```diff
@@ -145,6 +149,17 @@ func (receipt *Receipt) AddTransactionLogs(logs ...*TransactionLog) *Receipt {
...
 }
 
+// ExecutionRevertMsg returns the list of execution revert error logs stored in receipt.
+func (receipt *Receipt) ExecutionRevertMsg() string {
+	return receipt.executionRevertMsg
+}
+
+// AddExecutionRevertMsg add executionerrorlogs to receipt and filter out "" log.
+func (receipt *Receipt) AddExecutionRevertMsg(revertReason string) *Receipt {
+	receipt.executionRevertMsg = revertReason
```
Comment (dustinxie) [inline action/receipt.go:159]: add condition  if receipt.executionRevertMsg == ""?

File: ioctl/cmd/action/actionhash.go [hunk 6]
```diff
@@ -23,14 +23,13 @@ import (
...
-	"github.com/iotexproject/iotex-proto/golang/iotexapi"
-	"github.com/iotexproject/iotex-proto/golang/iotextypes"
-
 	"github.com/iotexproject/iotex-core/action/protocol/staking"
 	"github.com/iotexproject/iotex-core/ioctl/cmd/alias"
 	"github.com/iotexproject/iotex-core/ioctl/config"
 	"github.com/iotexproject/iotex-core/ioctl/output"
 	"github.com/iotexproject/iotex-core/ioctl/util"
+	"github.com/iotexproject/iotex-proto/golang/iotexapi"
+	"github.com/iotexproject/iotex-proto/golang/iotextypes"
```
Comment (dustinxie) [inline ioctl/cmd/action/actionhash.go:32]: no need to move

File: action/protocol/execution/evm/evmstatedbadapter.go (hunk 2 above)
Comment (koseoyoung) [reply action/protocol/execution/evm/evmstatedbadapter.go:30]: I will revert back, some lint automatically moved these -

File: action/receipt_test.go [hunk 7]
```diff
@@ -76,7 +77,8 @@ func TestSerDer(t *testing.T) {
 	require.Equal(receipt.ContractAddress, receipt2.ContractAddress)
 
 	hash := receipt.Hash()
-	require.Equal("9b1d77d8b8902e8d4e662e7cd07d8a74179e032f030d92441ca7fba1ca68e0f4", hex.EncodeToString(hash[:]))
+	oldHash := "9b1d77d8b8902e8d4e662e7cd07d8a74179e032f030d92441ca7fba1ca68e0f4"
+	require.Equal(oldHash, hex.EncodeToString(hash[:]))
```
Comment (dustinxie) [inline action/receipt_test.go:81]: add one more test, like
```go
// starting HawaiiHeight execution revert message is added to receipt
receipt.ExecutionRevertMsg = "test"
hash = receipt.Hash()
r.NotEqual(oldHash, h)
```

File: go.mod [hunk 8]
```diff
@@ -23,7 +23,7 @@ require (
 	github.com/iotexproject/iotex-address v0.2.2
 	github.com/iotexproject/iotex-antenna-go/v2 v2.4.0
 	github.com/iotexproject/iotex-election v0.3.5-0.20201031050050-c3ab4f339a54
-	github.com/iotexproject/iotex-proto v0.4.4-0.20201029172022-a8466422b0f1
+	github.com/iotexproject/iotex-proto v0.4.6-0.20201214081026-5a986f015db3
```
Comment (dustinxie) [inline go.mod:26]: i released your check-in as v0.4.6, you can update it

File: action/protocol/execution/evm/evm.go [hunk 9]
```diff
@@ -233,6 +238,16 @@ func ExecuteContract(
 		hu.IsPre(config.Greenland, blkCtx.BlockHeight) && receipt.Status == uint64(iotextypes.ReceiptStatus_ErrCodeStoreOutOfGas) {
 		receipt.AddTransactionLogs(stateDB.TransactionLogs()...)
 	}
+
+	if hu.IsPost(config.Hawaii, blkCtx.BlockHeight) && receipt.Status == uint64(iotextypes.ReceiptStatus_ErrExecutionReverted) {
+		if retval != nil && bytes.Equal(retval[:4], revertSelector) {
```
Comment (dustinxie) [inline action/protocol/execution/evm/evm.go:243]: can we combine these 2 if into a single one?

File: action/receipt.go (hunk 5 above)
Comment (koseoyoung) [reply action/receipt.go:159]: what do you want to handle in this condition?

**Review by CoderZhi — APPROVED** [review]

File: action/receipt_test.go [hunk 10]
```diff
@@ -76,7 +77,13 @@ func TestSerDer(t *testing.T) {
...
+
+	// starting HawaiiHeight execution revert message is added to receipt
+	receipt.executionRevertMsg = "test"
+	hash2 := receipt.Hash()
+	require.NotEqual(oldHash, hash2)
```
Comment (dustinxie) [inline action/receipt_test.go:86]: hex.EncodeToString(hash2)

Comment (koseoyoung) [reply action/receipt_test.go:86]: good catch! thx

File: action/receipt.go (hunk 5 above)
Comment (dustinxie) [reply action/receipt.go:160]: i mean
```go
if receipt.executionRevertMsg == "" {
   receipt.executionRevertMsg = revertReason
}
```
so it captures the first error string, and won't be overwritten on later calls, thought that's not the case (AddExecutionRevertMsg is only called once in your code change)

Comment (dustinxie) [reply action/receipt.go:160]: and better 
```go
if receipt.executionRevertMsg == "" &&  revertReason != "" {
   receipt.executionRevertMsg = revertReason
}
```

**Review by dustinxie — APPROVED** [review]

File: action/receipt.go (hunk 5 above)
Comment (koseoyoung) [reply action/receipt.go:160]: thanks for the clarification!

**Pull request merged**